presence_channel.go

```diff
@@ -0,0 +1,249 @@
+	"errors"
+	"fmt"
+)
+
+// Member represents a channel member.
+type Member struct {
+	ID   string
+	Info json.RawMessage
+}
+
+// PresenceChannel provides information about the users that are currently
+// subscribed.
+//
+// Note: Bind() does not fire pusher:member_added/removed, use
+// BindMemberAdded/Removed() instead.
+type PresenceChannel interface {
+	Channel
+	// BindMemberAdded returns a channel that receives a Member value when a user
+	// joins the channel.
+	BindMemberAdded() chan Member
+	// BindMemberRemoved returns a channel that receives a user ID when a user
+	// leaves the channel.
+	BindMemberRemoved() chan string
+	// Members returns the member info of each user currently subscribed to the
+	// channel.
+	Members() map[string]Member
+	// Member returns a representation of the channel member with the given ID.
+	// If the member isn't in the channel, the returned bool is `false`.
+	Member(id string) (bool, Member)
+	// Me returns the member for the current user.
+	//
+	// Possible errors:
+	//  - not subscribed - subscription must succeed before calling Me()
+	//  - invalid channel data - pusher server provided invalid JSON
+	//  - missing member for current user - pusher server violated protocol
+	Me() (*Member, error)
+	// MemberCount returns the number of users connected to the channel.
+	MemberCount() int
+}
+
+// presenceChannel implements the internalChannel and PresenceChannel interfaces
+type presenceChannel struct {
+	*privateChannel
+
+	memberAdded   chan Member
+	memberRemoved chan string
+	members       map[string]Member
+}
+
+func newPresenceChannel(baseChannel *channel) *presenceChannel {
+	privateChannel := &privateChannel{channel: baseChannel}
+	return &presenceChannel{
+		privateChannel: privateChannel,
+		memberAdded:    make(chan Member),
+		memberRemoved:  make(chan string),
+		members:        map[string]Member{},
+	}
+}
+
+// presenceChannelSubscriptionData is sent from Pusher to the client in the
+// subscription_succeeded event.
+// https://pusher.com/docs/channels/library_auth_reference/pusher-websockets-protocol/#pusher_internalsubscription_succeeded-pusher-channels-greater-client
+type presenceChannelSubscriptionData struct {
+	Presence struct {
+		IDs   []string                   `json:"ids"`
+		Hash  map[string]json.RawMessage `json:"hash"`
+		Count int                        `json:"count"`
+	} `json:"presence"`
+}
+
+type presenceChannelData struct {
+	UserID string `json:"user_id"`
+}
+
+type presenceChannelMemberAddedData struct {
+	UserID   string          `json:"user_id"`
+	UserInfo json.RawMessage `json:"user_info"`
+}
+
+type presenceChannelMemberRemovedData struct {
+	UserID string `json:"user_id"`
+}
+
+func (c *presenceChannel) handleEvent(event string, data json.RawMessage) {
+	if event == pusherInternalMemberAdded {
+		var member presenceChannelMemberAddedData
+		err := UnmarshalDataString(data, &member)
+		if err != nil {
+			c.client.sendError(err)
+		} else {
+			c.mutex.Lock()
```

Comment:
`listen` and `handleEvent` are using different mutexes, defined on the `Client` and `channel` structs respectively. Because they guard different things, there shouldn't be a deadlock here.

Reply:
Ah yep that totally makes sense. It tripped me up that the calls look the same but refer to different `c`s 😁 . Thanks for clarifying!